Android login and signup

kornelijepetak opened the pull request:
Closes #1132, #1133 

Following things are not implemented and should/will be implemented in separate PRs (as agreed with @willsb):

* Binding back button and FAB to appropriate VM props so that they get enabled or disabled accordingly (since we are using native Android behaviour and not our own buttons, this is a bit more complicated than on iOS) #1207 
* Google login is currently not implemented. Only the text label is there (which currently does nothing) #1208 
* Currently the back button goes a step back to be consistent with our VM. Only on the first step does back button lead again to the onboarding view. Nevertheless I have checked the point on the issue as completed. If this is not the desired behaviour, @pe0ny, please advise!

We have decided not to extract styles for the views for now. We will refactor if needed in future for components to be reusable. But we don't have many of those ATM.

willsb commented on Toggl.Giskard/Activities/LoginActivity.cs:
```diff
@@ -1,18 +1,41 @@
-{
-    [Activity(Theme = "@style/Theme.AppCompat.Light.DarkActionBar"), MvxActivityPresentation]
+{  
+    [Activity(Theme="@style/Theme.AppCompat.Light.NoActionBar", WindowSoftInputMode=SoftInput.AdjustResize)]
```
Add spaces before and after the `=`

willsb commented on Toggl.Giskard/Activities/LoginActivity.cs:
```diff
@@ -1,18 +1,41 @@
+        private void navigationClick(object sender, NavigationClickEventArgs args)
+        {
+            if (ViewModel.IsLoading)
+                return;
```
Early return on same line

willsb commented on Toggl.Giskard/Activities/MainActivity.cs:
```diff
@@ -1,11 +1,13 @@
 {
-    [Activity(Label = "View for MainViewModel")]
-    public class MainActivity : MvxActivity
+    [Activity(Theme="@style/Theme.AppCompat.Light.NoActionBar")]
```
Please add `[MvxActivityPresentation]`

willsb commented on Toggl.Giskard/Resources/layout/LoginActivity.axml:
```diff
@@ -2,19 +2,95 @@
+        android:id="@+id/LoginToolbar"
+        android:layout_width="match_parent"
+        android:layout_height="?attr/actionBarSize"
+        android:background="?attr/colorPrimary"
```
I don't think we should default to the primary color. Since our app won't follow the classic Material design approach for colors, you should instead use a color resource here.

willsb commented on Toggl.Giskard/Resources/layout/LoginActivity.axml:
```diff
@@ -2,19 +2,95 @@
+        android:inputType="textEmailAddress"
+        android:layout_height="30dp"
+        android:id="@+id/EmailTextField"
+        android:paddingLeft="16dp"
```
I like those to be in sorted by length in ascending order with `layout_` attributes at the end, regardless of their length. Yes, the nit is picked

willsb commented on Toggl.Giskard/Resources/layout/LoginActivity.axml:
```diff
@@ -2,19 +2,95 @@
+        android:layout_marginBottom="17dp"
+        android:layout_marginLeft="16dp" />
     <android.support.design.widget.FloatingActionButton
+        local:MvxBind="Click NextCommand; Enabled NextIsEnabled"
```
Please remove this binding, since it's not doing anything

willsb commented on Toggl.Giskard/Resources/values/Strings.xml:
```diff
@@ -1,7 +1,8 @@
+    <string name="Login">Log in</string>
     <string name="AppName">Toggl</string>
     <string name="EmailAddress">Email address</string>
+    <string name="LoginTrouble">Sorry, we\’re having trouble logging you in. Please try again.</string>
```
Where are you even using this?

willsb commented on Toggl.PrimeRadiant.Realm/packages.config:
```diff
@@ -0,0 +1,15 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
No `packages.config` files, please

kornelijepetak commented:
@willsb I have changed what you requested. One thing I changed partially, because I don't think it makes sense :)

Ordering items by string length in XML is OK, but just as you want layout elements to be at the end, I want the `id` and the `MvxBind` on top because these are the most important attributes of an element and I'd like those to be on top. `id` is short so it gets on top anyway, but `MvxBind` is the second most important thing (i.e. the purpose of the element), so yeah. We can argue more over it if you'd like, but I think this is more reasonable than just enforcing length ordering (which serves only aesthetic purposes). `Nitpicking denied!` :)

Other than that, I have added a settings screen to this PR with a simple back and logout button to be able to logout of the app. This is required to prevent constant uninstall to be able to test the login again. Since the project is still in its starting phase, I wouldn't want to wast much more time to submit it as a separate PR. Hopefully, you will agree.

kornelijepetak commented:
Since Signup (#1133) is almost a spin off the Login, I have integrated it with this PR. It will really take 1 more minute to review. Since it uses almost all the same components as the Login, it would be very superfluous to separate it. I am sorry if I have caused any inconvenience.

pe0ny commented:
> Currently the back button goes a step back to be consistent with our VM. Only on the first step does back button lead again to the onboarding view. Nevertheless I have checked the point on the issue as completed. If this is not the desired behaviour, @pe0ny, please advise!

LGTM 👍

kornelijepetak commented:
Ok, in the end, Settings view was removed to make PR cleaner.
Focusing will be solved as a separate issue/PR: #1229

willsb commented on Toggl.Giskard/Activities/LoginActivity.cs:
```diff
@@ -1,19 +1,56 @@
+
+        public override bool OnKeyDown(Keycode keyCode, KeyEvent e)
+        {
+            if(keyCode == Keycode.Back)
```
Since you are not calling base here, I think one won't be able to change the device's ringer volume while logging in 😄

willsb replied:
While at it, please use braces here, early return and then call `base.OnKeyDown` for all the other events

willsb commented on Toggl.Giskard/Activities/LoginActivity.cs:
```diff
@@ -1,19 +1,56 @@
+        }
+
+        private void navigationClick(object sender, NavigationClickEventArgs args) 
+            => executeBackCommand();
```
We as a team don't like expression bodied void methods

kornelijepetak replied:
I will change this, but there's several of those accross the codebase (some of them even blamed to you) :)
Not sure if we don't like them. :)

Why not though, if they are this simple? We do it a lot in setters.

willsb commented on Toggl.Giskard/Activities/LoginActivity.cs:
```diff
@@ -1,19 +1,56 @@
+        private void navigationClick(object sender, NavigationClickEventArgs args) 
+            => executeBackCommand();
+
+        private void executeBackCommand() { 
```
We are not building pyramids here, so no egyptians please

willsb commented on Toggl.Giskard/Resources/layout/SplashScreen.axml:
```diff
@@ -4,7 +4,12 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent">
     <TextView
```
Why are you even changing this file?

kornelijepetak replied:
By accident of course. I am just not sure how this happened at all.
